integration/route-ranking-test.ts

```diff
@@ -0,0 +1,162 @@
+import { PassThrough } from "node:stream";
+import type { Page } from "@playwright/test";
+import { test, expect } from "@playwright/test";
+
+import {
+  createAppFixture,
+  createFixture,
+  js,
+} from "./helpers/create-fixture.js";
+import { PlaywrightFixture } from "./helpers/playwright-fixture.js";
+
+let ROOT_FILE_CONTENTS = js`
+  import { Outlet, Scripts } from "@remix-run/react";
+
+  export default function App() {
+    return (
+      <html lang="en">
+        <body>
+          <Outlet />
+          <Scripts />
+        </body>
+      </html>
+    );
+  }
+`;
+
+async function setup(files: { [key: string]: string }, page: Page) {
+  let buildStdio = new PassThrough();
+  let buildOutput: string;
+
+  let fixture = await createFixture({
+    buildStdio,
+    files,
+  });
+
+  let chunks: Buffer[] = [];
+  buildOutput = await new Promise<string>((resolve, reject) => {
+    buildStdio.on("data", (chunk) => chunks.push(Buffer.from(chunk)));
+    buildStdio.on("error", (err) => reject(err));
+    buildStdio.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
+  });
+
+  let appFixture = await createAppFixture(fixture);
+  console.log("buildOutput\n", buildOutput);
+  let app = new PlaywrightFixture(appFixture, page);
+  return { fixture, appFixture, app, buildOutput };
+}
+
+test.describe("route ranking", () => {
+  test("chooses the first matching file in the case of ties", async ({
+    page,
+  }) => {
+    let { app } = await setup(
+      {
+        "app/root.tsx": ROOT_FILE_CONTENTS,
+        "app/routes/($locale).apple-pay.jsx": js`
+          import { useLoaderData } from '@remix-run/react';
+
+          export function loader() {
+            return ":locale?/apple-pay"
+          }
+
+          export default function Component() {
+            let data = useLoaderData();
+            return (
+              <>
+                <h1>:locale?/apple-pay</h1>
+                <pre>{data}</pre>
+              </>
+            );
+          }
+        `,
+        "app/routes/($locale).legal.$article.jsx": js`
+          import { useLoaderData } from '@remix-run/react';
+
+          export function loader() {
+            return ":locale?/legal/:article"
+          }
+
+          export default function Component() {
+            let data = useLoaderData();
+            return (
+              <>
+                <h1>:locale?/legal/:article</h1>
+                <pre>{data}</pre>
+              </>
+            );
+          }
+        `,
+      },
+      page
+    );
+
+    await app.goto("/legal/apple-pay");
+    await page.waitForSelector("h1");
+    expect(await app.getHtml("h1")).toMatch(":locale?/apple-pay");
+    expect(await app.getHtml("pre")).toMatch(":locale?/apple-pay");
```

Comment:
The tie here occurs because `($locale).apple-pay.jsx` and `($locale).legal.$article.jsx` explode out to the following in the react router routes:

```
:locale/apple-pay
/apple-pay
:locale/legal/:article
legal/:article
```

And then the URL `/legal/apple-pay` matches both of these routes with an identical score:

```
:locale/apple-pay
legal/:article
```

In this first test, we expect `($locale).apple-pay.jsx` to win due to alphabetical ordering